From 211eccc0447f4259cbdf9579786865ee8e3657ff Mon Sep 17 00:00:00 2001
From: Ira Weiny <ira.weiny@intel.com>
Date: Wed, 11 Mar 2015 01:41:35 -0400
Subject: [PATCH] infiniband-diags/rdma-ndd: add --pidfile option

rdma-ndd start up script works better with a pidfile.  This was only partially
implemented in the start up script and not at all in the daemon.

Reported-by: Dan Ben Yosef <danby@mellanox.com>
Signed-off-by: Ira Weiny <ira.weiny@intel.com>

[ Dropped the Makefile.am hunk to avoid having to re-run automake.
  Creating /run is redundant anyway. -- mschmidt ]
---
diff --git a/doc/man/rdma-ndd.8.in b/doc/man/rdma-ndd.8.in
index 1a7a1bb9b5..f0a542d08f 100644
--- a/doc/man/rdma-ndd.8.in
+++ b/doc/man/rdma-ndd.8.in
@@ -64,6 +64,9 @@ Number of times to attempt to retry setting of the node description on failure.
 .sp
 \fB\-\-foreground, \-f\fP
 Run in the foreground instead of as a daemon
+.sp
+\fB\-\-pidfile <pidfile>\fP
+specify a pid file (daemon mode only)
 .SS Configuration flags
 .\" Define the common option -z
 .
diff --git a/doc/rst/rdma-ndd.8.in.rst b/doc/rst/rdma-ndd.8.in.rst
index afe7470c6f..708a5ab0e8 100644
--- a/doc/rst/rdma-ndd.8.in.rst
+++ b/doc/rst/rdma-ndd.8.in.rst
@@ -58,6 +58,9 @@ Number of times to attempt to retry setting of the node description on failure.
 **--foreground, -f**
 Run in the foreground instead of as a daemon
 
+**--pidfile <pidfile>**
+specify a pid file (daemon mode only)
+
 
 Configuration flags
 -------------------
diff --git a/etc/rdma-ndd.init.in b/etc/rdma-ndd.init.in
index 24ce1bdb1c..75e6d370d0 100644
--- a/etc/rdma-ndd.init.in
+++ b/etc/rdma-ndd.init.in
@@ -75,7 +75,7 @@ start () {
 	return 1
     fi
     echo -n "Starting RDMA Node Description Daemon: "
-    @sbindir@/rdma-ndd > /dev/null
+    @sbindir@/rdma-ndd --pidfile $pidfile > /dev/null
     RETVAL=$?
     if [[ $RETVAL -eq 0 ]]; then
         success
diff --git a/src/rdma-ndd.c b/src/rdma-ndd.c
index bbde6af8de..194c2cd865 100644
--- a/src/rdma-ndd.c
+++ b/src/rdma-ndd.c
@@ -68,6 +68,7 @@ char *sys_dir = DEF_SYS_DIR;
 int failure_retry_rate = DEFAULT_RETRY_RATE;
 int set_retry_cnt = DEFAULT_RETRY_COUNT;
 int foreground = 0;
+char *pidfile = NULL;
 
 static void newline_to_null(char *str)
 {
@@ -205,6 +206,9 @@ static int process_opts(void *context, int ch, char *optarg)
 {
 	unsigned long tmp;
 	switch (ch) {
+	case 0:
+		pidfile = optarg;
+		break;
 	case 'f':
 		foreground = 1;
 		break;
@@ -352,6 +356,29 @@ static void monitor(void)
 	}
 }
 
+static void remove_pidfile(void)
+{
+        if (pidfile)
+		unlink(pidfile);
+}
+
+static void write_pidfile(void)
+{
+	FILE *f;
+	if (pidfile) {
+		remove_pidfile();
+		f = fopen(pidfile, "w");
+		if (f) {
+			fprintf(f, "%d\n", getpid());
+			fclose(f);
+		} else {
+			syslog(LOG_ERR, "Failed to write pidfile : %s\n",
+				pidfile);
+			exit(errno);
+		}
+	}
+}
+
 int main(int argc, char *argv[])
 {
 	int fd;
@@ -368,6 +395,7 @@ int main(int argc, char *argv[])
 			"Number of times to attempt to retry setting "
 			"of the node description on failure\n"},
 		{"foreground", 'f', 0, NULL, "run in the foreground instead of as a daemon\n"},
+		{"pidfile", 0, 1, "<pidfile>", "specify a pid file (daemon mode only)\n"},
 		{0}
 	};
 
@@ -377,8 +405,6 @@ int main(int argc, char *argv[])
 	if (!ibd_nd_format)
 		ibd_nd_format = DEFAULT_ND_FORMAT;
 
-	setup_udev();
-
 	if (!foreground) {
 		closelog();
 		openlog("rdma-ndd", LOG_PID, LOG_DAEMON);
@@ -386,8 +412,11 @@ int main(int argc, char *argv[])
 			syslog(LOG_ERR, "Failed to daemonize\n");
 			exit(errno);
 		}
+		write_pidfile();
 	}
 
+	setup_udev();
+
 	syslog(LOG_INFO, "Node Descriptor format (%s)\n", ibd_nd_format);
 
 	fd = open(SYS_HOSTNAME, O_RDONLY);
@@ -398,5 +427,7 @@ int main(int argc, char *argv[])
 
 	monitor();
 
+	remove_pidfile();
+
 	return 0;
 }
-- 
2.4.3

